submissions/m-ruslan/js-oop-practice/app.js

```diff
@@ -0,0 +1,146 @@
+
+  // **if enemy hits the player
+  if (
+    ((this.x + enemyPadding <= player.x + playerPadding &&
+      this.x + colStep - enemyPadding >= player.x + playerPadding) ||
+      (this.x + enemyPadding >= player.x + playerPadding &&
+        this.x + enemyPadding <= player.x + colStep - playerPadding)) &&
+    Math.round(this.y / 10, 0) === Math.round(player.y / 10, 0)
+  ) {
+    winsBalance -= 1;
+    winsBalEl.innerHTML = winsBalance;
+    prevTryResult.innerHTML = "<em>You lost previous time<em>";
+    player.x = pDefualtXPos;
+    player.y = pDefualtYPos;
+  }
+
+  // **restarts enemy's running
+  if (this.x > colStep * 6) {
+    this.x = -colStep;
+  }
+};
+
+// Draw the enemy on the screen, required method for game
+Enemy.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+// Now write your own player class
+// This class requires an update(), render() and
+// a handleInput() method.
+var Player = function (x, y) {
+  this.sprite = "images/char-boy.png";
+  this.x = x;
+  this.y = y;
+  console.log("Player position:", this.x, this.y);
+};
+Player.prototype.update = function () {
+  // **if player reaches the water (wins)
+  if (this.y === -30) {
+    winsBalance += 1;
+    winsBalEl.innerHTML = winsBalance;
+    prevTryResult.innerHTML = "<strong>You won previous time<strong>";
+    this.x = pDefualtXPos;
+    this.y = pDefualtYPos;
+  }
+};
+
+Player.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Player.prototype.handleInput = function (direction) {
+  var newPosX, newPosY;
+  var maxY = rowStep * 5 - 30;
+  var minY = -30;
+  var maxX = colStep * 4;
+  var minX = 0;
+
+  switch (direction) {
+    case "left":
+      newPosX = this.x - colStep;
+      this.x = newPosX < minX ? this.x : newPosX;
+      break;
+    case "right":
+      newPosX = this.x + colStep;
+      this.x = newPosX > maxX ? this.x : newPosX;
+      break;
+    case "up":
+      newPosY = this.y - rowStep;
+      this.y = newPosY < minY ? this.y : newPosY;
+      break;
+    case "down":
+      newPosY = this.y + rowStep;
+      this.y = newPosY > maxY ? this.y : newPosY;
+      break;
+    default:
+      break;
+  }
+};
+
+// Now instantiate your objects.
+// Place all enemy objects in an array called allEnemies
+// Place the player object in a variable called player
+var rowStep = 83;
+var colStep = 101;
+var playerPadding = 17;
+var enemyPadding = 2;
+var pDefualtXPos = colStep * 2;
+var pDefualtYPos = rowStep * 5 - 30;
+
+var player = new Player(pDefualtXPos, pDefualtYPos);
+
+var enemy1 = new Enemy(-colStep, rowStep * 3 - 30, 100);
+var enemy2 = new Enemy(-colStep, rowStep * 2 - 30, 200);
+var enemy3 = new Enemy(-colStep, rowStep * 1 - 30, 300);
+var enemy4 = new Enemy(-1000 - colStep, rowStep * 3 - 30, 200);
+var enemy5 = new Enemy(-1000 - colStep, rowStep * 2 - 30, 250);
+var enemy6 = new Enemy(-1000 - colStep, rowStep * 1 - 30, 300);
```

Comment:
Quite many repetitions. Can you devise an approach to benefit from calculable values, dependence on enemy index and eliminating repetitons?
Imagine we need to create dozens of enemies.

Reply:
@OleksiyRudenko thank you
I'll refactor my code and will come back to you for one more review)